### PR TITLE
interpolation variable

### PR DESCRIPTION
**Pull request by tormath1**
In this PR: 

I added a function `Interpolate` with two args: the HCL writer (since there's no interpolation with state AFAIK) and an interpolation map which map a value with its interpolation name. (`<resource_type>.<resource_name>.(id|self_link)`)

Feel free to test it. 

example of output: 
```tf
security_groups = ["sg-8a5124ec", "sg-01514a64", "${aws_security_group.Xiskc.id}", "${aws_security_group.inuev.id}"]
``` 

:warning: This is TF0.11 format (https://github.com/cycloidio/terracognita/issues/82) :warning: 

Close https://github.com/cycloidio/terracognita/issues/37

### CODE REVIEW

File: provider/import.go [hunk 1]
```diff
@@ -130,6 +134,16 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
 						return errors.Wrapf(err, "error while calculating the satate of resource %q", t)
 					}
 				}
+				// we construct a map[string]string to perform
+				// interpolation later. By default, the value to interpolate is the ID
+				// but in some case (ex: google), it can be different.
```
Comment (xlr-8) [inline provider/import.go]: Different how?

File: provider/import.go [hunk 2]
```diff
@@ -139,10 +153,15 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
...
+			err := util.Interpolate(hcl.(*h.Writer), interpolation)
+			if err != nil {
+				return errors.Wrapf(err, "error while Interpolate Config")
+			}
+		}
 		fmt.Fprintf(out, "\rWriting HCL ...")
 		logger.Log("msg", "writing the HCL")
-
-		err := hcl.Sync()
+		err = hcl.Sync()
```
Comment (xlr-8) [inline provider/import.go]: But why? :stuck_out_tongue:

File: provider/resource.go [hunk 3]
```diff
@@ -35,6 +35,13 @@ type Resource interface {
 	// Type is the type of resource (ex: aws_instance)
 	Type() string
 
+	// InstanceState is the Terraform state of the resource
+	// it contains important elements likes `Attributes`
```
Comment (xlr-8) [inline provider/resource.go]: like*

File: provider/import.go (hunk 1 above)
Comment (tormath1) [reply provider/import.go]: actually the comment is misleading. I modified it consequently. 
> it can be different

resource key could be `self_link` with Google.

File: provider/import.go [hunk 4]
```diff
@@ -149,10 +153,15 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
 	}
 
 	if hcl != nil {
+		if _, ok := hcl.(*h.Writer); ok {
+			err := util.Interpolate(hcl.(*h.Writer), interpolation)
+			if err != nil {
+				return errors.Wrapf(err, "error while Interpolate Config")
+			}
+		}
```
Comment (xescugc) [inline provider/import.go]: When I was reading the `util.Interpolate` I already thought that it should go to the `hcl` pkg, and now even more so the implementation is the one that tries to convert to the specific type even with that I don't like that as the Interface was ment to avoid knowing the implementation behind it, and this brakes it.

Ways to solve this:
* Have the `util` be the one that converts so it expects a `write.Writer` not an `hcl.Writer` (still not good IMO)
* Move it to `hcl` pkg
* Add it to the `write.Wiriter` interface
* Make the `Sync()` expect the `map[string]string`.

The last 2 are the ones I prefer as they delay the implementation to the actual one (State will do nothing as no Interpolation is needed and HCL will do something).

File: provider/import.go [hunk 5]
```diff
@@ -176,3 +185,13 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
...
 }
+
+func getKey(s *terraform.InstanceState) string {
+	// google special case
+	// sometimes, the resource is referred by its self_link (https://...)
+	// so we need to identify if we reference the ressource by its ID or its selfLink (the key)
+	if _, ok := s.Attributes["self_link"]; ok {
+		return "self_link"
+	}
+	return "id"
```
Comment (xescugc) [inline provider/import.go]: I've commented it to @tormath1 on PM that this could be much more things than just the `id`, on an `aws_instance` the relations could be done with https://www.terraform.io/docs/providers/aws/r/instance.html#attributes-reference any of those. But maybe for now only the ID it's the important one.

File: provider/import.go [hunk 6]
```diff
@@ -65,6 +67,8 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
 	fmt.Fprintf(out, "Importing with filters: %s", f)
 	logger.Log("filters", f.String())
 
+	interpolation := make(map[string]string)
```
Comment (xescugc) [inline provider/import.go]: Add documentations on what it holds and why is it used.

File: provider/import.go [hunk 7]
```diff
@@ -130,6 +130,16 @@ func Import(ctx context.Context, p Provider, hcl, tfstate writer.Writer, f *filt
...
+				// we construct a map[string]string to perform
+				// interpolation later. By default, the value to interpolate is the ID
+				// but in some case (ex: google), it can be different.
+				state := r.InstanceState()
+				if state != nil {
+					key := getKey(state)
+					value := state.Attributes[key]
+					interpolatedValue := fmt.Sprintf("${%s.%s.%s}", r.Type(), r.Name(), key)
+					interpolation[value] = interpolatedValue
+				}
```
Comment (xescugc) [inline provider/import.go]: ~Also all this logic could be done on the `r.HCL` for example as it has all the logic and the actual scope for all this information (state and Name and all of them)~

Maybe better have a method on the `Resource` that actually returns all the possible interpolated values (right now just ID but potentially more) as the Resource has access to the schema and could potentially know them.

But maybe for now it's ok this way.

**Comment by tormath1** [comment]
@xescugc ready for review !

File: provider/import.go (hunk 5 above)
Comment (xlr-8) [reply provider/import.go]: Among the list we want for sure - and that we already widely use (some services I can think of):
`id` (general reference), `arn` (iam,alb,etc), `name` (iam,r53),  `fqdn` (r53)

Comment (tormath1) [reply provider/import.go]: Well, my current thought would be to have an `AttributesReference` associated to a `Resource`. This attribute would be simply feed by https://github.com/cycloidio/tfdocs.

Interpolation would now be made there: 
https://github.com/cycloidio/terracognita/blob/5709caa05628bc41ad1d1a49b462da93bfa47a5d/provider/import.go#L92 since we have all the `resources`.

Then looping on the resources, if a value of resources.AttributesReferences[*] match with a value in the `resources`, we interpolate.
WDYT ?

Comment (tormath1) [reply provider/import.go]: Well, my current thought would be to have an `AttributesReference` associated to a `Resource`. This attribute would be simply feed by https://github.com/cycloidio/tfdocs.

Interpolation would now be made there: 
https://github.com/cycloidio/terracognita/blob/5709caa05628bc41ad1d1a49b462da93bfa47a5d/provider/import.go#L92 since we have all the `resources`.

Then looping on the resources, if a value of resources.AttributesReferences[*] match with a value in the `resources`, we interpolate.
WDYT ?

**Comment by tormath1** [comment]
@xescugc ready for review. 

In the latest commit, I totally modified the was I was updating the resources. At the beginning I was trying to simply `bytes.ReplaceAll` but it's clearly not enough, since we are going to replace also custom tags (e.g `cycloid.io` was always interpolated), it's also not enough to avoid interpolation with the resource itself (let's suppose you are inside `aws_instance.toto` you can't refer to `aws_instance.toto.some_field` from this resource) or cyclic interpolation.

To summarize: 

* we are now using `tfdocs`: https://github.com/cycloidio/terracognita/pull/81/commits/e7234db5a5b063d23a348b7bfc1f146cb01d86a2 in order to get all potentials `key` to be interpolated.

* we are walking recursively into a `resources` interface in order to interpolate the `value` whom are able to be interpolated (no self reference and no cyclic reference.)

I tried out with a large infra (security groups and instances) and it works as expected.

File: hcl/writer.go [hunk 8]
```diff
@@ -128,30 +129,72 @@ func (w *Writer) Sync() error {
...
+	for rt, resource := range resources.(map[string]map[string]interface{}) {
+		// we loop over a resource (e.g: aws_instance.oDSOj)
+		for name, block := range resource {
+			src := reflect.ValueOf(block)
+			// this will store the updated block
+			dest := reflect.New(src.Type()).Elem()
+			// walk through the resources to interpolate the good values
+			walk(dest, src, i, name, rt)
+			// remove reflect.Value wrapper from dest
+			resources.(map[string]map[string]interface{})[rt][name] = dest.Interface()
```
Comment (xescugc) [inline hcl/writer.go]: Could you space up this part a bit? Comment+Code+\n so it's easy to read if not it's too dense.

File: hcl/writer.go [hunk 9]
```diff
@@ -128,30 +129,72 @@ func (w *Writer) Sync() error {
...
 		}
 	}
 	return nil
 }
+
+// walk through a resource block. it's easier since we do not how the block is made
+func walk(dest, src reflect.Value, interpolate map[string]string, name string, resourceType string) {
```
Comment (xescugc) [inline hcl/writer.go]: Improve the docs, what does it actually do? It goes though all the `src` and put it on `dest`, so instead of "returning" the result is the "dest". (better wording ofc hehe).

File: hcl/writer.go [hunk 10]
```diff
@@ -128,30 +129,72 @@ func (w *Writer) Sync() error {
...
+		for iter.Next() {
+			// New gives us a pointer, but again we want the value
+			destValue := reflect.New(iter.Value().Type()).Elem()
+			walk(destValue, iter.Value(), interpolate, name, resourceType)
+			dest.SetMapIndex(iter.Key(), destValue)
+		}
+
+	// what we want to interpolate is a string
+	// we do not interpolate a custom tag (like cycloid.io) since it's key.
+	case reflect.String:
```
Comment (xescugc) [inline hcl/writer.go]: Commented on PM with you already but to keep the record, could it be anything else than String no? Int, bool etc.

File: provider/resource.go [hunk 11]
```diff
@@ -119,6 +126,24 @@ func NewResource(id, rt string, p Provider) Resource {
...
+	var (
+		res = new(tfdocs.Resource)
+		err error
+	)
+	switch r.provider.String() {
+	case "aws":
+		res, err = awsdocs.GetResource(r.resourceType)
+	case "google":
+		res, err = googledocs.GetResource(r.resourceType)
+	}
```
Comment (xescugc) [inline provider/resource.go]: Return the errors first, basically check them after the end of the `switch`

Comment (xescugc) [inline provider/resource.go]: This could also be done by having an static map where the key is the `r.provider.String()` and the value `func (r string) (*tfdocresource.Resource, error)`, this way is an static map, and easy to add more instead of having to add a new `case` to the switch.

Comment (tormath1) [reply provider/resource.go]: @xescugc good catch ! it's way cleaner. Ready for final review :wink:

File: hcl/writer.go [hunk 12]
```diff
@@ -123,3 +124,83 @@ func (w *Writer) Sync() error {
...
+			dest := reflect.New(src.Type()).Elem()
+
+			// walk through the resources to interpolate the good values
+			walk(dest, src, i, name, rt)
+
+			// remove reflect.Value wrapper from dest
+			resources.(map[string]map[string]interface{})[rt][name] = dest.Interface()
+		}
+	}
+	return nil
```
Comment (xlr-8) [inline hcl/writer.go]: Why return an error, if it never will?

File: hcl/writer.go [hunk 13]
```diff
@@ -123,3 +124,83 @@ func (w *Writer) Sync() error {
...
+}
+
+// walk through a resource block. it's easier since we do not how the block is made
```
Comment (xlr-8) [inline hcl/writer.go]: do not know how* ?

File: provider/resource.go [hunk 14]
```diff
@@ -108,6 +115,14 @@ type resource struct {
...
 var (
 	autogeneratedAWSResourcesRe = regexp.MustCompile(`^aws:(?:autoscaling|cloudformation)`)
+	providerResources           = map[string]func(string) (*tfdocs.Resource, error){
+		"aws": func(rt string) (*tfdocs.Resource, error) {
+			return awsdocs.GetResource(rt)
+		},
+		"google": func(rt string) (*tfdocs.Resource, error) {
+			return googledocs.GetResource(rt)
+		},
+	}
```
Comment (xescugc) [inline provider/resource.go]: Ok a better way:

* Define a function type `func(rt string) (*tfdocs.Resource, error)`
* Then the map just have to have `"google" googledocs.GetResource`

Which make is much more simpler (that was my first intention but IDK why I read it wrong on the implementation).

Or actually the way you have it now you can still directly use the format without defining the function actually :smile:

File: hcl/writer.go (hunk 12 above)
Comment (xescugc) [reply hcl/writer.go]: Yeh it could be removed now as it's using a different logic.

**Comment by tormath1** [comment]
@xescugc final rebase has been made :)

File: provider/resource.go [hunk 15]
```diff
@@ -119,6 +132,21 @@ func NewResource(id, rt string, p Provider) Resource {
 	}
 }
 
+func (r *resource) AttributesReference() ([]string, error) {
+	resourceFunc := providerResources[r.provider.String()]
```
Comment (xescugc) [inline provider/resource.go]: Check the `ok` also, as it could not exists.

Comment (tormath1) [reply provider/resource.go]: @xescugc done in 6fbac852e750d05ed7db8fbf972ea43f5e522ead :)
